streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java

```diff
@@ -189,38 +190,52 @@ private void prepareClose(final boolean clean) {
     @Override
     public void closeClean(final Map<TopicPartition, Long> checkpoint) {
         Objects.requireNonNull(checkpoint);
-        close(true);
+        close(true, false);
 
         log.info("Closed clean");
     }
 
     @Override
     public void closeDirty() {
-        close(false);
+        close(false, false);
 
         log.info("Closed dirty");
     }
 
-    private void close(final boolean clean) {
+    @Override
+    public void closeAndRecycleState() {
+        prepareClose(true);
+        close(true, true);
```

Comment:
nit: I'd suggest we have a separate `recycle()` function rather than piggy-backing on `close` with additional flag, since the `clean` is always true, i.e. we would always just 1) write checkpoints, 2) call stateMgr.recycle(), 3) then transit to CLOSED and if it is not in CREATED / RUNNING then do nothing. Though it has some duplicated lines but it is a bit straight-forward.

Reply:
I was debating this a lot actually. Ultimately I decided to reuse the existing `close` methods because I was worried we might change or fix something in `close` and forget to add it to `closeAndRecycle`.  Since `closeAndRecycle` should always be a subset of the actions taken in `close` I thought this would be more future proof...WDYT?

Reply:
Well, I think it still makes sense to have `closeDirty` and `closeClean` fall back to the same `close` method since the only difference is whether to rethrow exceptions, whereas `closeAndRecycle` actually has different contents.
But I guess for that reason right there it makes sense to duplicate only the relevant parts of `close`.